src/sentry/snuba/outcomes.py

```diff
@@ -0,0 +1,330 @@
+
+from sentry.snuba.sessions_v2 import (
+    get_constrained_date_range,
+    massage_sessions_result,
+    SimpleGroupBy,
+    InvalidField,
+)
+from sentry.utils.outcomes import Outcome
+from datetime import datetime
+from sentry.search.utils import InvalidQuery
+
+
+"""
+The new Outcomes API defines a "metrics"-like interface which is can be used in
+a similar way to "sessions" and "discover"
+# "Metrics"
+We have 2 "metrics" that we can query:
+
+- `quantity` (counter): The relevant stat based on category:
+        standard event: # of outcomes
+        session: number of events within the session
+        attachment: quantity in bytes of attachments
+- `times_seen` (counter): The number of outcomes that occurred --
+        differs from quantity in that:
+            it would count # of attachments, and sessions
+            if we ever aggregate multiple events into one outcome,
+            this would be used to get the number of outcomes
+
+# Tags / Grouping
+
+The Outcomes data can be grouped / filtered
+by these fields
+- `project`
+- `outcome`
+- `reason`
+- `category`
+"""
+
+
+class Field(ABC):
+    @abstractmethod
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def extract_from_row(self, row: Dict[str, Any], group: Optional[Dict[str, Any]] = None) -> int:
+        raise NotImplementedError()
+
+    @abstractmethod
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        raise NotImplementedError()
+
+
+class QuantityField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        return ["quantity"]
+
+    def extract_from_row(self, row: Dict[str, Any], group: Optional[Dict[str, Any]] = None) -> int:
+        if row is None:
+            return 0
+        return int(row["quantity"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        return ["sum", "quantity", "quantity"]
+
+
+class TimesSeenField(Field):
+    def get_snuba_columns(self, raw_groupby: Optional[List[str]] = None) -> List[str]:
+        return ["times_seen"]
+
+    def extract_from_row(self, row: Dict[str, Any], group: Optional[Dict[str, Any]] = None) -> int:
+        if row is None:
+            return 0
+        return int(row["times_seen"])
+
+    def aggregation(self, dataset: Dataset) -> List[str]:
+        if dataset == Dataset.Outcomes:
+            return ["sum", "times_seen", "times_seen"]
+        else:
+            # RawOutcomes doesnt have times_seen, do a count instead
+            return ["count()", "", "times_seen"]
+
+
+class Dimension(SimpleGroupBy, ABC):
+    @abstractmethod
+    def resolve_filter(self, raw_filter: List[str]) -> List[DataCategory]:
+        """
+        Based on the input filter, map it back to the clickhouse representation
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def map_row(self, row: Dict[str, Any]) -> None:
```

Comment:
Here, since `row` is modified by the function `Dict` or `MutableMapping` works.

Reply:
👍🏼